cpp/tests/streams/replace_test.cpp

```diff
@@ -104,9 +104,9 @@ TEST_F(ReplaceTest, NormalizeNansAndZeros)
 
 TEST_F(ReplaceTest, NormalizeNansAndZerosMutable)
 {
-  auto nan          = std::numeric_limits<double>::quiet_NaN();
-  auto input_column = cudf::test::make_type_param_vector<double>({-0.0, 0.0, -nan, nan, nan});
-  cudf::test::fixed_width_column_wrapper<double> input(input_column.begin(), input_column.end());
-  cudf::mutable_column_view mutable_view = cudf::column(input, cudf::test::get_default_stream());
-  cudf::normalize_nans_and_zeros(mutable_view, cudf::test::get_default_stream());
+  auto nan   = std::numeric_limits<double>::quiet_NaN();
+  auto data  = cudf::test::make_type_param_vector<double>({-0.0, 0.0, -nan, nan, nan});
+  auto input = cudf::test::fixed_width_column_wrapper<double>(data.begin(), data.end()).release();
+  auto view  = input->mutable_view();
+  cudf::normalize_nans_and_zeros(view, cudf::test::get_default_stream());
```

Comment:
would not mind this option, but nothing wrong with current code either
```suggestion
  cudf::normalize_nans_and_zeros(input->mutable_view(), cudf::test::get_default_stream());
```

Reply:
@davidwendt this is what I was getting at above. My interpretation of 
> The view must be created from a non-temporary column and also must be non-temporary itself so that it is not implicitly converted to a column_view.

was that this change would somehow break things. Perhaps I was misunderstanding though.

Reply:
That is exactly what happens. The temporary created here
```
cudf::normalize_nans_and_zeros(input->mutable_view(), cudf::test::get_default_stream());
```
causes the compiler to call the `column_view const&` API instead of the `mutable_column_view&` API.
The current code insures the appropriate API is called.

Reply:
The original change made here: https://github.com/rapidsai/cudf/pull/17436 was an attempt correct the API call by creating a `mutable_column_view` variable but inadvertently created the view to a destroyed temp column.

Reply:
How does that happen? I wouldn't have expected that overload to ever be selected in this way unless the overload for the same type was actually impossible to call...

Reply:
Again, I'm guessing because it is a temporary and passing a non-const temporary usually makes no sense since any modifications the object that occur int function are just thrown away. I suppose the compiler is trying hard to help here by finding a _better_ API candidate to call.
I feel like this https://godbolt.org/z/hW7cnKPxW illustrates that as well.

Perhaps we should have an implicit operator conversion from `mutable_column_view` to `column_view`
I would not expect that to be a common thing and making it explicit not be a big deal in our code base.

Reply:
> I feel like this https://godbolt.org/z/hW7cnKPxW illustrates that as well.

OK I put together [a slightly modified version of your example](https://godbolt.org/z/Ynxn6W784) that helped me. I found your example a bit different since there is no overload of the function that actually accepts an instance of `hello` itself. I would have thought that would always be preferred.

Reply:
Before the changes in #17436, we were passing a `mutable_column_view` rvalue to `cudf::normalize_nans_and_zeros`. Since we cannot bind a rvalue to a non-const lvalue reference, the `cudf::normalize_nans_and_zeros(mutable_column_view &)` overload could not be called, and the compiler instead converted the `mutable_column_view` to `column_view` so that the overload (`cudf::normalize_nans_and_zeros(column_view const &)`) with the const reference parameter could be invoked.
However, while trying to create a `mutable_column_view` lvalue, I accidentally created the view to a rvalue which does not make sense. Thank you for the fix, @davidwendt!

Reply:
> Since we cannot bind a rvalue to a non-const lvalue reference

I forgot that this was a rule, thanks for stating it out explicitly. I guess the compiler prevents this since there's no sensible reason to allow this and it protects against user error modifying a parameter in a way that would have no effect.